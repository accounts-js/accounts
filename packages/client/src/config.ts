import {
  config as sharedConfig,
  AccountsCommonConfiguration,
  HookListener,
  PasswordSignupFields,
} from '@accounts/common';
import { Store, Middleware } from 'redux';

export interface TokenStorage {
  getItem(key: string): Promise<string>;
  removeItem(key: string): Promise<string>;
  setItem(key: string, value: string): Promise<string>;
}

export type AccountsClientConfiguration = AccountsCommonConfiguration & {
  store?: Store<object>;
  reduxLogger?: Middleware;
  reduxStoreKey?: string;
  tokenStorage?: TokenStorage;
  server?: string;
  tokenStoragePrefix?: string;
  title?: string;
  requestPermissions?: any[];
  requestOfflineToken?: object;
  forceApprovalPrompt?: object;
  requireEmailVerification?: boolean;
  loginPath?: string;
  signUpPath?: string;
  resetPasswordPath?: string;
  profilePath?: string;
  changePasswordPath?: string;
  homePath?: string;
  signOutPath?: string;
  onEnrollAccountHook?: HookListener;
  onResetPasswordHook?: HookListener;
  onVerifyEmailHook?: HookListener;
  onSignedInHook?: HookListener;
  onSignedOutHook?: HookListener;
  onResumedSessionHook?: HookListener;
  onUserCreated?: (user: object) => Promise<any>;
  loginOnSignUp?: boolean;
  history?: string[];
  persistImpersonation?: boolean;
};

const localStorageToTokenStorage = () => {
  return {
    getItem(key: string) {
      return Promise.resolve(localStorage.getItem(key));
    },
    removeItem(key: string) {
      const value = localStorage.getItem(key);
      localStorage.removeItem(key);
      return Promise.resolve(value);
    },
    setItem(key: string, value: string) {
      localStorage.setItem(key, value);
      return Promise.resolve(value);
    },
  };
};

export default {
  ...sharedConfig,
  store: null,
  reduxLogger: null,
  reduxStoreKey: 'accounts',
  tokenStorage: localStorageToTokenStorage(),
  server: '',
  tokenStoragePrefix: '',
  title: '',
  requestPermissions: [],
  requestOfflineToken: {},
  forceApprovalPrompt: {},
  requireEmailVerification: false,
  loginPath: '/',
  signUpPath: null,
  resetPasswordPath: null,
  profilePath: '/',
  changePasswordPath: null,
  homePath: '/',
  signOutPath: '/',
  loginOnSignUp: true,
  persistImpersonation: true,
};
