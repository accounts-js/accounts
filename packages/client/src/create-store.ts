import { createStore, applyMiddleware, compose, Store, Middleware } from 'redux';
import { Map } from 'immutable';
import { combineReducers } from 'redux-immutable';

export default ({
  reducers = {},
  middleware = [],
  state = Map(),
}: {
  reducers?: object;
  middleware?: Middleware[];
  state?: Map<string, any>;
}) => createStore(combineReducers(reducers), state, compose(applyMiddleware(...middleware)));
