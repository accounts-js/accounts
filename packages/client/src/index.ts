export { AccountsClient } from './accounts-client';
export { AccountsClientOptions } from './types';
export { TransportInterface } from './transport-interface';
