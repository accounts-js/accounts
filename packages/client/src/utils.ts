import jwtDecode from 'jwt-decode';

interface JwtDecodeData {
  exp: number;
  iat: number;
}

export const isTokenExpired = (token: string): boolean => {
  const currentTime = Date.now() / 1000;
  const decodedToken = jwtDecode<JwtDecodeData>(token);
  return decodedToken.exp < currentTime;
};
