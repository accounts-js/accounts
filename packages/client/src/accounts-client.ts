import { LoginResult, MFALoginResult, Tokens, ImpersonationResult, User } from '@accounts/types';
import { TransportInterface } from './transport-interface';
import { TokenStorage, AccountsClientOptions } from './types';
import { tokenStorageLocal } from './token-storage-local';
import { isTokenExpired } from './utils';

enum TokenKey {
  AccessToken = 'accessToken',
  RefreshToken = 'refreshToken',
  MfaToken = 'mfaToken',
  OriginalAccessToken = 'originalAccessToken',
  OriginalRefreshToken = 'originalRefreshToken',
}

const defaultOptions = {
  tokenStorage: tokenStorageLocal,
  tokenStoragePrefix: 'accounts',
};

export class AccountsClient {
  public transport: TransportInterface;
  private options: AccountsClientOptions & typeof defaultOptions;
  private storage: TokenStorage;

  constructor(options: AccountsClientOptions, transport: TransportInterface) {
    this.options = { ...defaultOptions, ...options };
    this.storage = this.options.tokenStorage;

    if (!transport) {
      throw new Error('A valid transport is required');
    }
    this.transport = transport;
    this.transport.client = this;
  }

  /**
   * Get the tokens from the storage
   */
  public async getTokens(original?: boolean): Promise<Tokens | null> {
    const accessToken = await this.storage.getItem(
      original
        ? this.getTokenKey(TokenKey.OriginalAccessToken)
        : this.getTokenKey(TokenKey.AccessToken)
    );
    const refreshToken = await this.storage.getItem(
      original
        ? this.getTokenKey(TokenKey.OriginalRefreshToken)
        : this.getTokenKey(TokenKey.RefreshToken)
    );
    if (!accessToken || !refreshToken) {
      return null;
    }
    return { accessToken, refreshToken };
  }

  /**
   * Store the tokens in the storage
   */
  public async setTokens(tokens: Tokens, original?: boolean): Promise<void> {
    await this.storage.setItem(
      original
        ? this.getTokenKey(TokenKey.OriginalAccessToken)
        : this.getTokenKey(TokenKey.AccessToken),
      tokens.accessToken
    );
    await this.storage.setItem(
      original
        ? this.getTokenKey(TokenKey.OriginalRefreshToken)
        : this.getTokenKey(TokenKey.RefreshToken),
      tokens.refreshToken
    );
  }

  /**
   * Remove the tokens from the storage
   */
  public async clearTokens(original?: boolean): Promise<void> {
    await this.storage.removeItem(
      original
        ? this.getTokenKey(TokenKey.OriginalAccessToken)
        : this.getTokenKey(TokenKey.AccessToken)
    );
    await this.storage.removeItem(
      original
        ? this.getTokenKey(TokenKey.OriginalRefreshToken)
        : this.getTokenKey(TokenKey.RefreshToken)
    );
  }

  /**
   * Authenticate the user with a specific service (not creating a session)
   */
  public async authenticateWithService(
    service: string,
    credentials: { [key: string]: any }
  ): Promise<boolean> {
    return this.transport.authenticateWithService(service, credentials);
  }

  /**
   * Login the user with a specific service
   */
  public async loginWithService(
    service: string,
    credentials: { [key: string]: any }
  ): Promise<LoginResult | Omit<MFALoginResult, 'mfaToken'>> {
    const response = await this.transport.loginWithService(service, credentials);

    if ((response as LoginResult).tokens) {
      await this.setTokens((response as LoginResult).tokens);
    } else {
      await this.storage.setItem(
        this.getTokenKey(TokenKey.MfaToken),
        (response as MFALoginResult).mfaToken
      );
    }

    return response;
  }

  /**
   * Performs the mfa needed challenge and logs in afterwards
   */
  public async performMfaChallenge(challenge: string, params: any): Promise<LoginResult> {
    const mfaToken = await this.storage.getItem(this.getTokenKey(TokenKey.MfaToken));

    if (!mfaToken) {
      throw new Error('mfaToken is not available in storage');
    }

    const loginToken = await this.transport.performMfaChallenge(challenge, mfaToken, params);

    const result = await this.loginWithService('mfa', { mfaToken, loginToken });

    await this.storage.removeItem(this.getTokenKey(TokenKey.MfaToken));

    return result as any;
  }

  /**
   * Refresh the user session
   * If the tokens have expired try to refresh them
   */
  public async refreshSession(force = false): Promise<Tokens | null> {
    const tokens = await this.getTokens();
    if (tokens) {
      try {
        const isAccessTokenExpired = isTokenExpired(tokens.accessToken);
        const isRefreshTokenExpired = isTokenExpired(tokens.refreshToken);
        // See if accessToken is expired and refreshToken is not
        if ((force || isAccessTokenExpired) && !isRefreshTokenExpired) {
          // Request a new token pair
          const refreshedSession = await this.transport.refreshTokens(
            tokens.accessToken,
            tokens.refreshToken
          );

          await this.setTokens(refreshedSession.tokens);
          return refreshedSession.tokens;
        } else if (isRefreshTokenExpired) {
          // Refresh token is expired, user must sign back in
          await this.clearTokens();
          return null;
        }
        return tokens;
      } catch (err) {
        await this.clearTokens();
        throw err;
      }
    }
    return null;
  }

  /**
   * Impersonate to another user.
   */
  public async impersonate(impersonated: {
    userId?: string;
    username?: string;
    email?: string;
  }): Promise<ImpersonationResult> {
    const tokens = await this.getTokens();

    if (!tokens) {
      throw new Error('An access token is required');
    }

    const res = await this.transport.impersonate(tokens.accessToken, impersonated);

    if (!res.authorized || !res.tokens) {
      throw new Error(`User unauthorized to impersonate`);
    } else {
      await this.setTokens(tokens, true);
      await this.setTokens(res.tokens);
      return res;
    }
  }

  /**
   * Stop the user impersonation.
   */
  public async stopImpersonation(): Promise<void> {
    const tokens = await this.getTokens(true);
    if (tokens) {
      await this.setTokens(tokens);
      await this.clearTokens(true);
    }
  }

  /**
   * Get the user infos
   */
  public async getUser(): Promise<User> {
    return this.transport.getUser();
  }

  /**
   * Logout the user
   * Call the server to invalidate the tokens
   * Clean user local storage
   */
  public async logout(): Promise<void> {
    const tokens = await this.getTokens();
    if (tokens) {
      await this.transport.logout();
    }
    await this.clearTokens();
  }

  private getTokenKey(tokenName: TokenKey): string {
    return `${this.options.tokenStoragePrefix}:${tokenName}`;
  }
}
