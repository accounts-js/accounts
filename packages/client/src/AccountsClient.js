// @flow

import { isFunction, isString, has } from 'lodash';
import type { Map } from 'immutable';
import type { Store } from 'redux';
import jwtDecode from 'jwt-decode';
import {
  AccountsError,
  validators,
  CreateUserType,
  PasswordLoginUserType,
  LoginReturnType,
  UserObjectType,
  TokensType,
} from '@accounts/common';
import config from './config';
import createStore from './createStore';
import reducer, { loggingIn, setUser, clearUser } from './module';
import type { TransportInterface } from './TransportInterface';

const isValidUserObject = (user: PasswordLoginUserType) => has(user, 'user') || has(user, 'email') || has(user, 'id');

const ACCESS_TOKEN = 'accounts:accessToken';
const REFRESH_TOKEN = 'accounts:refreshToken';

const getTokenKey = (type: string, options: Object) =>
  (isString(options.localStoragePrefix) && options.localStoragePrefix.length > 0 ? `${options.localStoragePrefix}:${type}` : type);

export class AccountsClient {
  options: Object
  transport: TransportInterface
  store: Store<Map<string, any>>

  constructor(options: Object, transport: TransportInterface) {
    this.options = options;
    if (!transport) {
      throw new AccountsError('A REST or GraphQL transport is required');
    }

    this.transport = transport;

    const middleware = options.reduxLogger ? [
      options.reduxLogger,
    ] : [];

    this.store = createStore({
      reducers: {
        accounts: reducer,
      },
      middleware,
    });
  }
  getState(): Map<string, any> {
    return this.store.getState().get('accounts');
  }
  user(): UserObjectType {
    return this.getState().get('user').toJS();
  }
  tokens(): TokensType {
    return {
      accessToken: localStorage.getItem(getTokenKey(ACCESS_TOKEN, this.options)),
      refreshToken: localStorage.getItem(getTokenKey(REFRESH_TOKEN, this.options)),
    };
  }
  clearTokens() {
    localStorage.removeItem(getTokenKey(ACCESS_TOKEN, this.options));
    localStorage.removeItem(getTokenKey(REFRESH_TOKEN, this.options));
  }
  clearUser() {
    this.store.dispatch(clearUser());
  }
  resumeSession(): Promise<void> {
    // TODO Should there be any additional resume session logic here?
    return this.refreshSession();
  }
  async refreshSession(): Promise<void> {
    const { accessToken, refreshToken } = this.tokens();
    if (accessToken && refreshToken) {
      try {
        const decodedRefreshToken = jwtDecode(refreshToken);
        const currentTime = Date.now() / 1000;
        // Refresh token is expired, user must sign back in
        if (decodedRefreshToken.exp < currentTime) {
          this.clearTokens();
          this.clearUser();
        } else {
          // Request a new token pair
          const refreshedSession : LoginReturnType =
            await this.transport.refreshTokens(accessToken, refreshToken);
          localStorage.setItem(
            getTokenKey(ACCESS_TOKEN, this.options),
            refreshedSession.tokens.accessToken,
          );
          localStorage.setItem(
            getTokenKey(REFRESH_TOKEN, this.options),
            refreshedSession.tokens.refreshToken);
          this.store.dispatch(setUser(refreshedSession.user));
        }
      } catch (err) {
        this.clearTokens();
        this.clearUser();
        throw new AccountsError('falsy token provided');
      }
    } else {
      this.clearTokens();
      this.clearUser();
    }
  }
  async createUser(user: CreateUserType, callback: ?Function): Promise<void> {
    if (!user || user.password === undefined) {
      throw new AccountsError(
        'Unrecognized options for create user request',
        {
          username: user && user.username,
          email: user && user.email,
        },
        400,
      );
    }

    if (!validators.validatePassword(user.password)) {
      throw new AccountsError('Password is required');
    }

    if (!validators.validateUsername(user.username) && !validators.validateEmail(user.email)) {
      throw new AccountsError('Username or Email is required');
    }

    try {
      const userId = await this.transport.createUser(user);
      if (callback && isFunction(callback)) {
        callback();
      }
      await this.loginWithPassword({ id: userId }, user.password);
    } catch (err) {
      if (callback && isFunction(callback)) {
        callback(err);
      }
      throw new AccountsError(err.message);
    }
  }

  async loginWithPassword(user: PasswordLoginUserType,
                          password: string,
                          callback: ?Function): Promise<void> {
    if (!password || !user) {
      throw new AccountsError('Unrecognized options for login request', user, 400);
    }
    if ((!isString(user) && !isValidUserObject(user)) || !isString(password)) {
      throw new AccountsError('Match failed', user, 400);
    }

    this.store.dispatch(loggingIn(true));
    try {
      const res : LoginReturnType = await this.transport.loginWithPassword(user, password);
      localStorage.setItem(getTokenKey(ACCESS_TOKEN, this.options), res.tokens.accessToken);
      localStorage.setItem(getTokenKey(REFRESH_TOKEN, this.options), res.tokens.refreshToken);
      this.store.dispatch(setUser(res.user));
      this.options.onSignedInHook();
      if (callback && isFunction(callback)) {
        callback();
      }
    } catch (err) {
      if (callback && isFunction(callback)) {
        callback(err);
      }
      throw new AccountsError(err.message);
    }
    this.store.dispatch(loggingIn(false), user);
  }
  loggingIn(): boolean {
    return (this.getState().get('loggingIn'): boolean);
  }
  isLoading(): boolean {
    return (this.getState().get('isLoading'): boolean);
  }
  async logout(callback: ?Function): Promise<void> {
    try {
      const { accessToken } = this.tokens();
      await this.transport.logout(accessToken);
      this.clearTokens();
      this.store.dispatch(clearUser());
      if (callback && isFunction(callback)) {
        callback();
      }
      this.options.onSignedOutHook();
    } catch (err) {
      if (callback && isFunction(callback)) {
        callback(err);
      }
      throw new AccountsError(err.message);
    }
  }
}

const Accounts = {
  instance: AccountsClient,
  ui: {

  },
  config(options: Object, transport: TransportInterface) {
    this.instance = new AccountsClient({
      ...config,
      ...options,
    }, transport);
  },
  user(): UserObjectType {
    return this.instance.user();
  },
  options(): Object {
    return this.instance.options;
  },
  createUser(user: CreateUserType, callback: ?Function): Promise<void> {
    return this.instance.createUser(user, callback);
  },
  loginWithPassword(user: PasswordLoginUserType,
                    password: string,
                    callback: Function): Promise<void> {
    return this.instance.loginWithPassword(user, password, callback);
  },
  loggingIn(): boolean {
    return this.instance.loggingIn();
  },
  isLoading(): boolean {
    return this.instance.isLoading();
  },
  logout(callback: ?Function): Promise<void> {
    return this.instance.logout(callback);
  },
  tokens(): TokensType {
    return this.instance.tokens();
  },
  resumeSession(): Promise<void> {
    return this.instance.resumeSession();
  },
  refreshSession(): Promise<void> {
    return this.instance.refreshSession();
  },
};

export default Accounts;

// TODO Could this be handled better?
// if (typeof window !== 'undefined') {
//   window.onload = async () => {
//     if (Accounts.instance && Accounts.instance.resumeSession) {
//       await Accounts.resumeSession();
//     }
//   };
// }
