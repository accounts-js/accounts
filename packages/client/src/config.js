// @flow

import { config as sharedConfig } from '@accounts/common';
import type { AccountsCommonConfiguration } from '@accounts/common';
import type { Store } from 'redux';
import AccountsClient from './AccountsClient';
import redirect from './redirect';

export interface TokenStorage {
  getItem(key: string): Promise<string>,
  removeItem(key: string): Promise<string>,
  setItem(key: string, value: string): Promise<string>
}

export type AccountsClientConfiguration = AccountsCommonConfiguration & {
  store?: ?Store<Object, Object>,
  reduxLogger?: ?Object,
  reduxStoreKey?: string,
  tokenStorage?: ?TokenStorage,
  server?: string,
  tokenStoragePrefix?: string,
  title?: string,
  requestPermissions?: Array<any>,
  requestOfflineToken?: Object,
  forceApprovalPrompt?: Object,
  requireEmailVerification?: boolean,
  loginPath?: string,
  signUpPath?: ?string,
  resetPasswordPath?: ?string,
  profilePath?: string,
  changePasswordPath?: ?string,
  homePath?: string,
  signOutPath?: string,
  onEnrollAccountHook?: Function,
  onResetPasswordHook?: Function,
  onVerifyEmailHook?: Function,
  onSignedInHook?: Function,
  onSignedOutHook?: Function,
  onResumedSessionHook?: Function,
  onUserCreated?: (user: ?Object) => Promise<any>,
  loginOnSignUp?: boolean,
  history?: Object,
  persistImpersonation?: boolean
};

export default {
  ...sharedConfig,
  store: null,
  reduxLogger: null,
  reduxStoreKey: 'accounts',
  tokenStorage: global.localStorage,
  server: '',
  tokenStoragePrefix: '',
  title: '',
  requestPermissions: [],
  requestOfflineToken: {},
  forceApprovalPrompt: {},
  requireEmailVerification: false,
  loginPath: '/',
  signUpPath: null,
  resetPasswordPath: null,
  profilePath: '/',
  changePasswordPath: null,
  homePath: '/',
  signOutPath: '/',
  // TODO enable all of these
  // onSubmitHook: () => {},
  // onPreSignUpHook: () => new Promise(resolve => resolve()),
  // onPostSignUpHook: () => {},
  // onEnrollAccountHook: () => redirect(AccountsClient.options().loginPath || '/'),
  // onResetPasswordHook: () => redirect(AccountsClient.options().loginPath || '/'),
  // onVerifyEmailHook: () => redirect(AccountsClient.options().profilePath || '/'),
  // onSignedInHook: (accountsClient) => redirect(accountsClient.options.homePath || '/'),
  // onSignedOutHook: () => redirect(AccountsClient.options().signOutPath || '/'),
  loginOnSignUp: true,
  persistImpersonation: true,
};
