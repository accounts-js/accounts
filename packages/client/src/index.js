// @flow
import type { TokenStorage } from './config';
import type { TransportInterface } from './TransportInterface';

import { AccountsClient } from './AccountsClient';
import config from './config';
import reducer from './module';
import validate from './validate';

export default AccountsClient;

export {
  AccountsClient,
  config,
  reducer,
  validate,
};

export type {
  TransportInterface,
  TokenStorage,
};
