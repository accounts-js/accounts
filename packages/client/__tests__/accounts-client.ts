import { isTokenExpired } from '../src/utils';
import { AccountsClient } from '../src/accounts-client';

jest.mock('../src/utils');

const loggedInResponse = {
  sessionId: '1',
  tokens: {
    accessToken: 'accessToken',
    refreshToken: 'refreshToken',
  },
};

const impersonateResult = {
  authorized: true,
  tokens: { accessToken: 'newAccessToken', refreshToken: 'newRefreshToken' },
};

const tokens = {
  accessToken: 'accessTokenTest',
  refreshToken: 'refreshTokenTest',
};

const mockTransport = {
  loginWithService: jest.fn(() => Promise.resolve(loggedInResponse)),
  authenticateWithService: jest.fn(() => Promise.resolve(true)),
  logout: jest.fn(() => Promise.resolve()),
  refreshTokens: jest.fn(() => Promise.resolve(loggedInResponse)),
  sendResetPasswordEmail: jest.fn(() => Promise.resolve()),
  verifyEmail: jest.fn(() => Promise.resolve()),
  sendVerificationEmail: jest.fn(() => Promise.resolve()),
  impersonate: jest.fn(() => Promise.resolve(impersonateResult)),
};

describe('Accounts', () => {
  const accountsClient = new AccountsClient({}, mockTransport as any);

  afterEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

  it('requires a transport', async () => {
    try {
      new AccountsClient(null as any, null as any);
      throw new Error();
    } catch (err) {
      const { message } = err;
      expect(message).toEqual('A valid transport is required');
    }
  });

  describe('getTokens', () => {
    it('should return null when no tokens', async () => {
      const result = await accountsClient.getTokens();
      expect(result).toBe(null);
      expect(localStorage.getItem).toHaveBeenCalledTimes(2);
    });

    it('should return the tokens', async () => {
      await accountsClient.setTokens(tokens);
      const result = await accountsClient.getTokens();
      expect(tokens).toEqual(result);
      expect(localStorage.getItem).toHaveBeenCalledTimes(2);
    });

    it('should return the original tokens', async () => {
      await accountsClient.setTokens(tokens, true);
      const result = await accountsClient.getTokens(true);
      expect(tokens).toEqual(result);
      expect(localStorage.getItem).toHaveBeenCalledTimes(2);
    });
  });

  describe('setTokens', () => {
    it('should set the tokens', async () => {
      await accountsClient.setTokens(tokens);
      expect(localStorage.setItem).toHaveBeenCalledTimes(2);
      expect(localStorage.getItem('accounts:accessToken')).toEqual(tokens.accessToken);
      expect(localStorage.getItem('accounts:refreshToken')).toEqual(tokens.refreshToken);
    });

    it('should set the oiriginal tokens', async () => {
      await accountsClient.setTokens(tokens, true);
      expect(localStorage.setItem).toHaveBeenCalledTimes(2);
      expect(localStorage.getItem('accounts:originalAccessToken')).toEqual(tokens.accessToken);
      expect(localStorage.getItem('accounts:originalRefreshToken')).toEqual(tokens.refreshToken);
    });
  });

  describe('clearTokens', () => {
    it('should clear the tokens', async () => {
      await accountsClient.setTokens(tokens);
      await accountsClient.clearTokens();
      expect(localStorage.removeItem).toHaveBeenCalledTimes(2);
      expect(localStorage.getItem('accounts:accessToken')).toEqual(null);
      expect(localStorage.getItem('accounts:refreshToken')).toEqual(null);
    });

    it('should clear the oiriginal tokens', async () => {
      await accountsClient.setTokens(tokens, true);
      await accountsClient.clearTokens(true);
      expect(localStorage.removeItem).toHaveBeenCalledTimes(2);
      expect(localStorage.getItem('accounts:originalAccessToken')).toEqual(null);
      expect(localStorage.getItem('accounts:originalRefreshToken')).toEqual(null);
    });
  });

  describe('logout', () => {
    it('should clear the tokens', async () => {
      await accountsClient.logout();
      expect(localStorage.removeItem).toHaveBeenCalledTimes(2);
      expect(localStorage.getItem('accounts:accessToken')).toEqual(null);
      expect(localStorage.getItem('accounts:refreshToken')).toEqual(null);
    });

    it('should logout and clear the tokens', async () => {
      await accountsClient.setTokens(tokens);
      await accountsClient.logout();
      expect(localStorage.removeItem).toHaveBeenCalledTimes(2);
      expect(localStorage.getItem('accounts:accessToken')).toEqual(null);
      expect(localStorage.getItem('accounts:refreshToken')).toEqual(null);
      expect(mockTransport.logout).toHaveBeenCalled();
    });
  });

  describe('loginWithService', () => {
    it('calls transport', async () => {
      await accountsClient.loginWithService('password', {
        username: 'user',
        password: 'password',
      });
      expect(mockTransport.loginWithService).toHaveBeenCalledTimes(1);
      expect(mockTransport.loginWithService).toHaveBeenCalledWith('password', {
        username: 'user',
        password: 'password',
      });
    });

    it('set the tokens', async () => {
      await accountsClient.loginWithService('password', {
        username: 'user',
        password: 'password',
      });
      expect(localStorage.setItem).toHaveBeenCalledTimes(2);
      expect(localStorage.getItem('accounts:accessToken')).toEqual(
        loggedInResponse.tokens.accessToken
      );
      expect(localStorage.getItem('accounts:refreshToken')).toEqual(
        loggedInResponse.tokens.refreshToken
      );
    });
  });

  describe('authenticateWithService', () => {
    it('calls transport', async () => {
      await accountsClient.authenticateWithService('password', {
        username: 'user',
        password: 'password',
      });
      expect(mockTransport.authenticateWithService).toHaveBeenCalledTimes(1);
      expect(mockTransport.authenticateWithService).toHaveBeenCalledWith('password', {
        username: 'user',
        password: 'password',
      });
    });
  });

  describe('refreshSession', () => {
    it('should do nothing when no tokens', async () => {
      const result = await accountsClient.refreshSession();
      expect(result).not.toBeTruthy();
      expect(isTokenExpired).not.toHaveBeenCalled();
    });

    it('should call transport.refreshTokens if accessToken is expired and set the tokens', async () => {
      (isTokenExpired as jest.Mock).mockImplementationOnce(() => true);
      (isTokenExpired as jest.Mock).mockImplementationOnce(() => false);
      await accountsClient.setTokens(tokens);
      const result = await accountsClient.refreshSession();
      expect(result).toEqual(loggedInResponse.tokens);
      expect(isTokenExpired).toHaveBeenCalledWith(tokens.accessToken);
      expect(mockTransport.refreshTokens).toHaveBeenCalledWith(
        tokens.accessToken,
        tokens.refreshToken
      );
      expect(localStorage.setItem).toHaveBeenCalledTimes(4);
    });

    it('should do nothing if tokens are still valid', async () => {
      (isTokenExpired as jest.Mock).mockImplementation(() => false);
      await accountsClient.setTokens(tokens);
      const result = await accountsClient.refreshSession();
      expect(result).toEqual(tokens);
      expect(localStorage.setItem).toHaveBeenCalledTimes(2);
      expect(mockTransport.refreshTokens).not.toHaveBeenCalledWith();
    });

    it('should call transport.refreshTokens if force is required and set the tokens', async () => {
      (isTokenExpired as jest.Mock).mockImplementationOnce(() => false);
      (isTokenExpired as jest.Mock).mockImplementationOnce(() => false);
      await accountsClient.setTokens(tokens);
      const result = await accountsClient.refreshSession(true);
      expect(result).toEqual(loggedInResponse.tokens);
      expect(isTokenExpired).toHaveBeenCalledWith(tokens.accessToken);
      expect(mockTransport.refreshTokens).toHaveBeenCalledWith(
        tokens.accessToken,
        tokens.refreshToken
      );
      expect(localStorage.setItem).toHaveBeenCalledTimes(4);
    });

    it('should clear the tokens is refreshToken is expired', async () => {
      (isTokenExpired as jest.Mock).mockImplementationOnce(() => false);
      (isTokenExpired as jest.Mock).mockImplementationOnce(() => true);
      await accountsClient.setTokens(tokens);
      const result = await accountsClient.refreshSession();
      expect(result).toBe(null);
      expect(localStorage.removeItem).toHaveBeenCalledTimes(2);
      expect(mockTransport.refreshTokens).not.toHaveBeenCalledWith();
    });

    it('should clear the tokens and forward the error', async () => {
      try {
        (isTokenExpired as jest.Mock).mockImplementation(() => {
          throw new Error('Err');
        });
        await accountsClient.setTokens(tokens);
        await accountsClient.refreshSession();
        throw new Error();
      } catch (err) {
        expect(err.message).toBe('Err');
        expect(localStorage.removeItem).toHaveBeenCalledTimes(2);
        expect(mockTransport.refreshTokens).not.toHaveBeenCalledWith();
      }
    });
  });

  describe('impersonate', () => {
    it('should throw error if no tokens', async () => {
      try {
        await accountsClient.impersonate({ userId: 'test' });
        throw new Error();
      } catch (err) {
        expect(err.message).toBe('An access token is required');
      }
    });

    it('should throw if server return unauthorized', async () => {
      (mockTransport.impersonate as jest.Mock).mockImplementationOnce(() =>
        Promise.resolve({ authorized: false })
      );
      try {
        await accountsClient.setTokens(tokens);
        await accountsClient.impersonate({ userId: 'test' });
        throw new Error();
      } catch (err) {
        expect(err.message).toBe('User unauthorized to impersonate');
      }
    });

    it('should set tokens correctly if impersonation was authorized', async () => {
      await accountsClient.setTokens(tokens);
      await accountsClient.impersonate({ userId: 'test' });
      const userTokens = await accountsClient.getTokens();
      const originalTokens = await accountsClient.getTokens(true);
      expect(userTokens).toEqual(impersonateResult.tokens);
      expect(originalTokens).toEqual(tokens);
    });
  });

  describe('stopImpersonation', () => {
    it('should not replace tokens if not impersonating', async () => {
      await accountsClient.setTokens(tokens);
      await accountsClient.stopImpersonation();
      const userTokens = await accountsClient.getTokens();
      expect(userTokens).toEqual(tokens);
    });

    it('should replace the tokens', async () => {
      await accountsClient.setTokens(tokens);
      await accountsClient.setTokens(impersonateResult.tokens, true);
      await accountsClient.stopImpersonation();
      const userTokens = await accountsClient.getTokens();
      expect(userTokens).toEqual(impersonateResult.tokens);
    });
  });
});
