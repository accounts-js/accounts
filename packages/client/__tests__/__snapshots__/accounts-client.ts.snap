// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accounts performMfaChallenge throws error when no mfaToken exists in storage: mfaToken-not-available 1`] = `[Error: mfaToken is not available in storage]`;
